pkg/kubelet/volumemanager/reconciler/reconstruct_new.go

```diff
@@ -0,0 +1,243 @@
+	if readyToUnmount {
+		// Ensure devices that should be detached/unmounted are detached/unmounted.
+		rc.unmountDetachDevices()
+
+		// Clean up any orphan volumes that failed reconstruction.
+		rc.cleanOrphanVolumes()
+	}
+
+	if len(rc.volumesNeedDevicePath) != 0 {
+		rc.updateReconstructedDevicePaths()
+	}
+
+	if len(rc.volumesNeedReportedInUse) != 0 && rc.populatorHasAddedPods() {
+		// Once DSW is populated, mark all reconstructed as reported in node.status,
+		// so they can proceed with MountDevice / SetUp.
+		rc.desiredStateOfWorld.MarkVolumesReportedInUse(rc.volumesNeedReportedInUse)
+		rc.volumesNeedReportedInUse = nil
+	}
+}
+
+// readyToUnmount returns true when reconciler can start unmounting volumes.
+func (rc *reconciler) readyToUnmount() bool {
+	// During kubelet startup, all volumes present on disk are added as uncertain to ASW.
+	// Allow unmount only when DSW is fully populated to prevent unmounting volumes that
+	// did not reach DSW yet.
+	if !rc.populatorHasAddedPods() {
+		return false
+	}
+
+	// Allow unmount only when ASW device paths were corrected from node.status to prevent
+	// calling unmount with a wrong devicePath.
+	if len(rc.volumesNeedDevicePath) != 0 {
+		return false
+	}
+	return true
+}
+
+// reconstructVolumes tries to reconstruct the actual state of world by scanning all pods' volume
+// directories from the disk. For the volumes that cannot support or fail reconstruction, it will
+// put the volumes to volumesFailedReconstruction to be cleaned up later when DesiredStateOfWorld
+// is populated.
+func (rc *reconciler) reconstructVolumes() {
+	defer rc.updateLastSyncTime()
+	// Get volumes information by reading the pod's directory
+	podVolumes, err := getVolumesFromPodDir(rc.kubeletPodsDir)
+	if err != nil {
+		klog.ErrorS(err, "Cannot get volumes from disk, skip sync states for volume reconstruction")
+		return
+	}
+	reconstructedVolumes := make(map[v1.UniqueVolumeName]*globalVolumeInfo)
+	reconstructedVolumeNames := []v1.UniqueVolumeName{}
+	for _, volume := range podVolumes {
+		if rc.actualStateOfWorld.VolumeExistsWithSpecName(volume.podName, volume.volumeSpecName) {
+			klog.V(4).InfoS("Volume exists in actual state, skip cleaning up mounts", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+			// There is nothing to reconstruct
+			continue
+		}
+		reconstructedVolume, err := rc.reconstructVolume(volume)
+		if err != nil {
+			klog.InfoS("Could not construct volume information", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName, "err", err)
+			// We can't reconstruct the volume. Remember to check DSW after it's fully populated and force unmount the volume when it's orphaned.
+			rc.volumesFailedReconstruction = append(rc.volumesFailedReconstruction, volume)
+			continue
+		}
+		klog.V(4).InfoS("Adding reconstructed volume to actual state and node status", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
```

Comment:
use v(2)?

Reply:
This is actually copied from the old reconstruction, where it is at `V(4)`:
https://github.com/kubernetes/kubernetes/blob/3a79466ddde88e5161f1ca9754bebc62ef4ac8b9/pkg/kubelet/volumemanager/reconciler/reconstruct.go#L154.

Looking at logs, there is just V(2): `reconstruct_new.go:103] "Volume reconstruction finished"`. Since mount / unmount `MountVolume.TearDown succeeded for volume` is at `V(2)`, it makes sense to add reconstructed volumes there too. Added as a separate commit in both old/new reconstruction.